### PR TITLE
MQE: Skip DeduplicateAndMerge if possible

### PR DESCRIPTION
**Pull request by Konstantinov-Innokentii**
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+	"github.com/grafana/mimir/pkg/streamingpromql"
+	"github.com/grafana/mimir/pkg/streamingpromql/optimize/plan"
+	"github.com/grafana/mimir/pkg/streamingpromql/planning"
+	"github.com/grafana/mimir/pkg/streamingpromql/planning/core"
+	"github.com/grafana/mimir/pkg/streamingpromql/testutils"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+func TestEliminateDeduplicateAndMergeOptimizationPass(t *testing.T) {
```
Comment (charleskorn) [inline pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: Could you please add some tests for cases where we need to retain the `DeduplicateAndMerge` for cases with `or`, as well as vector/scalar comparisons with `bool` (eg. `metric == bool 6`)?

It would also be good to see tests for binary operations, to confirm that we're skipping them correctly.

File: pkg/streamingpromql/planning.go [hunk 2]
```diff
@@ -78,6 +78,11 @@ func NewQueryPlanner(opts EngineOpts) (*QueryPlanner, error) {
 		planner.RegisterQueryPlanOptimizationPass(plan.NewNarrowSelectorsOptimizationPass(opts.Logger))
 	}
 
+	// EliminateDeduplicateAndMergeO doesn't support delayed name removal yet.
```
Comment (charleskorn) [inline pkg/streamingpromql/planning.go]:
```suggestion
	// EliminateDeduplicateAndMerge doesn't support delayed name removal yet.
```

File: pkg/streamingpromql/planning.go [hunk 3]
```diff
@@ -78,6 +78,11 @@ func NewQueryPlanner(opts EngineOpts) (*QueryPlanner, error) {
...
+	if !opts.CommonOpts.EnableDelayedNameRemoval {
+		planner.RegisterQueryPlanOptimizationPass(plan.NewEliminateDeduplicateAndMergeOptimizationPass())
```
Comment (charleskorn) [inline pkg/streamingpromql/planning.go]: Should we run this before CSE? Then we don't have to think about how to handle deduplicated subexpressions in this optimisation pass.

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	if newRoot != nil {
+		plan.Root = newRoot
+	}
+	return plan, nil
+}
+
+func (e *EliminateDeduplicateAndMergeOptimizationPass) collectNodesToRemove(node planning.Node, parent planning.Node, childIndex int, nodesToRemove *[]dedupNodeInfo) {
+	// Binary operations are not supported yet.
+	// 1. It makes elimination logic much more complex - it's needed to track if name was dropped on both sides of the binary operation.
+	// 2. We likely can't do label projections when binary operation is involved, becaue of conflict error messages.
```
Comment (charleskorn) [inline pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge.go]: This isn't 100% correct: if the query contains aggregations within binary operations, for example `sum(rate(foo[5m])) / sum(rate(bar[5m]))`, then we can perform projections.

I think it's straightforward to remove `DeduplicateAndMerge` nodes provided they don't contain binary operations. For example, in `rate(foo[5m]) / rate(bar[5m])`, we can remove the `DeduplicateAndMerge` nodes around the `rate()` calls.

I also think we can remove `DeduplicateAndMerge` nodes that contain binary operations, but we'd need to inspect the binary operation to confirm if it can produce series with a `__name__` label, which sounds complex, as you say - would be good to write this down somewhere as a future improvement.

File: pkg/streamingpromql/planning.go (hunk 3 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/planning.go]: What's special about deduplicated subexpressions in this pass? I mean, isn't CSE just replaces subexpression with DuplicateConsumer, and that shouldn't affect the Deduplication?

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge.go (hunk 4 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge.go]: makes sense

File: pkg/streamingpromql/planning.go (hunk 3 above)
Comment (charleskorn) [reply pkg/streamingpromql/planning.go]: It may make the logic in `EliminateDeduplicateAndMergeOptimizationPass` more complex, as before CSE, the query plan is a tree, but afterwards, it may no longer be a tree due to multiple paths culminating in the same `Duplicate` node.

Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/planning.go]: Done, thanks for catching this.

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go (hunk 1 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: Addded

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge.go (hunk 4 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge.go]: So, I propose to keep it as it is and add support for `rate(foo[5m]) / rate(bar[5m])` expressions in the follow-up PR, just to push this over the line.

File: pkg/streamingpromql/planning.go [hunk 5]
```diff
@@ -65,6 +65,14 @@ func NewQueryPlanner(opts EngineOpts) (*QueryPlanner, error) {
 	planner.RegisterASTOptimizationPass(&ast.SortLabelsAndMatchers{}) // This is a prerequisite for other optimization passes such as common subexpression elimination.
 	// After query sharding is moved here, we want to move propagate matchers and reorder histogram aggregation here as well before query sharding.
 
+	// EliminateDeduplicateAndMerge doesn't support delayed name removal yet.
+	// This optimization pass is registered before CSE to keep the query plan as a simple tree structure.
+	// After CSE, the query plan may no longer be a tree due to multiple paths culminating in the same Duplicate node,
+	// which would make the elimination logic more complex.
+	if !opts.CommonOpts.EnableDelayedNameRemoval {
```
Comment (charleskorn) [inline pkg/streamingpromql/planning.go]: It'd be good to add a feature flag for this optimisation pass as well, so we can disable it easily if we discover any problems.

Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/planning.go]: Done

**Comment by Konstantinov-Innokentii** [comment]
Please, do not merge it yet, I think I found some corner cases.

**Comment by Konstantinov-Innokentii** [comment]
> Please, do not merge it yet, I think I found some corner cases.

So, my concern was that we are running upstream test cases with delayed name removal enabled, which lowers test coverage for the optimization, since it's runs only when delayed name removal is disabled :)
I addressed it in this commit, basically just adding more tests - https://github.com/grafana/mimir/pull/12528/commits/b1aa6fc6e15e418fc1243d9cd98a600c2c2f3a80.

File: cmd/mimir/config-descriptor.json [hunk 6]
```diff
@@ -2758,6 +2758,17 @@
               "fieldFlag": "querier.mimir-query-engine.enable-narrow-binary-selectors",
               "fieldType": "boolean",
               "fieldCategory": "experimental"
+            },
+            {
+              "kind": "field",
+              "name": "enable_eliminate_deduplicate_and_merge",
+              "required": false,
+              "desc": "Enable eliminating redundant DeduplicateAndMerge nodes from the query plan when it can be proven that each input series will produce a unique output series.",
```
Comment (tacole02) [inline cmd/mimir/config-descriptor.json]:
```suggestion
              "desc": "Enable eliminating redundant DeduplicateAndMerge nodes from the query plan when it can be proven that each input series produces a unique output series.",
```

File: operations/mimir/mimir-flags-defaults.json
Comment (charleskorn) [inline operations/mimir/mimir-flags-defaults.json:1]: Should `querier.mimir-query-engine.enable-eliminate-deduplicate-and-merge` be added here as well?

File: pkg/streamingpromql/planning.go [hunk 7]
```diff
@@ -65,6 +65,14 @@ func NewQueryPlanner(opts EngineOpts) (*QueryPlanner, error) {
...
 	// After query sharding is moved here, we want to move propagate matchers and reorder histogram aggregation here as well before query sharding.
 
+	// EliminateDeduplicateAndMerge doesn't support delayed name removal yet.
+	// This optimization pass is registered before CSE to keep the query plan as a simple tree structure.
+	// After CSE, the query plan may no longer be a tree due to multiple paths culminating in the same Duplicate node,
+	// which would make the elimination logic more complex.
+	if !opts.CommonOpts.EnableDelayedNameRemoval && opts.EnableEliminateDeduplicateAndMerge {
+		planner.RegisterQueryPlanOptimizationPass(plan.NewEliminateDeduplicateAndMergeOptimizationPass())
+	}
+
```
Comment (charleskorn) [inline pkg/streamingpromql/planning.go]: Rather than silently ignore the invalid configuration, I would return an error:

```suggestion
	// This optimization pass is registered before CSE to keep the query plan as a simple tree structure.
	// After CSE, the query plan may no longer be a tree due to multiple paths culminating in the same Duplicate node,
	// which would make the elimination logic more complex.
	if opts.EnableEliminateDeduplicateAndMerge {
		if opts.CommonOpts.EnableDelayedNameRemoval {
			return nil, errors.New("eliminating deduplicate and merge nodes is not supported with delayed name removal")
		}
	
		planner.RegisterQueryPlanOptimizationPass(plan.NewEliminateDeduplicateAndMergeOptimizationPass())
	}
```

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+	}{
+		"series with different labelset after name drop": {
+			data: `
+				load 1m
+					a{env="prod"} 0+1x10
+					a{env="test"} 0+2x10
+			`,
+			expr: `abs(rate(a[5m]))`,
+			expectedResult: `{env="prod"} => 0.016666666666666666 @[600000]
+{env="test"} => 0.03333333333333333 @[600000]`,
```
Comment (charleskorn) [inline pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: [nit] It might be easier to read `expectedResult` with slightly different formatting: (the extra whitespace can be removed with `testutils.TrimIndent` like we do for the query plan output in the tests above)

```suggestion
			expectedResult: `
				{env="prod"} => 0.016666666666666666 @[600000]
				{env="test"} => 0.03333333333333333 @[600000]
			`,
```

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+	if _, ok := node.(*core.DeduplicateAndMerge); ok {
+		count = 1
+	}
+	for _, child := range node.Children() {
+		count += countDeduplicateAndMergeNodes(child)
+	}
+	return count
+}
+
+func TestEliminateDeduplicateAndMergeOptimizationPass(t *testing.T) {
```
Comment (charleskorn) [inline pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: [nit]

```suggestion
func TestEliminateDeduplicateAndMergeOptimizationPassCorrectness(t *testing.T) {
```

Comment (charleskorn) [inline pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: One other idea for this: you could run all of our existing test cases (including the upstream ones) with the optimisation pass enabled to confirm there are no cases where this introduces an issue.

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go (hunk 8 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: Thanks!

File: operations/mimir/mimir-flags-defaults.json
Comment (Konstantinov-Innokentii) [reply operations/mimir/mimir-flags-defaults.json:1]: You mean in config-descriptor.json? It's already there

Comment (charleskorn) [reply operations/mimir/mimir-flags-defaults.json:1]: No, I mean in this file, `mimir-flags-defaults.json`.

Comment (Konstantinov-Innokentii) [reply operations/mimir/mimir-flags-defaults.json:1]: It seems only flag which are true by default are there for some reason. For example, there is no "enable-narrow-binary-selectors" there,

Comment (charleskorn) [reply operations/mimir/mimir-flags-defaults.json:1]: Right, makes sense. But why did `ruler.evaluation-consistency-max-delay` need to be added as part of this PR?

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go (hunk 9 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go]: Added TestEliminateDeduplicateAndMergeOptimizationDoesNotRegress, which runs upstream/ours test cases with delayed name removal disabled.

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+func TestEliminateDeduplicateAndMergeOptimizationDoesNotRegress(t *testing.T) {
+	runTestCasesWithDelayedNameRemovalDisabled(t, "upstream/*.test")
+	runTestCasesWithDelayedNameRemovalDisabled(t, "ours*/*.test")
+}
+
+func runTestCasesWithDelayedNameRemovalDisabled(t *testing.T, globPattern string) {
+
+	types.EnableManglingReturnedSlices = true
+	parser.ExperimentalDurationExpr = true
+	parser.EnableExperimentalFunctions = true
```
Comment (charleskorn) [inline pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go:574]: We should restore these to their previous values after the test finishes, to ensure this doesn't affect any other tests.

File: operations/mimir/mimir-flags-defaults.json
Comment (Konstantinov-Innokentii) [reply operations/mimir/mimir-flags-defaults.json:1]: It's generated automatically when you run `make reference-help` and it seems it was not added for some reason in the [PR](https://github.com/grafana/mimir/pull/12751) which added this setting

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go (hunk 10 above)
Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go:574]: Makes sense. But in the engine_test.go these are set in the init and not resetted later, should we fix it?
```
func init() {
	types.EnableManglingReturnedSlices = true
	parser.ExperimentalDurationExpr = true
	parser.EnableExperimentalFunctions = true

	// Set a tracer provider with in memory span exporter so we can check the spans later.
	otel.SetTracerProvider(
		tracesdk.NewTracerProvider(
			tracesdk.WithSpanProcessor(tracesdk.NewSimpleSpanProcessor(spanExporter)),
		),
	)
}
```

File: cmd/mimir/config-descriptor.json (hunk 6 above)
Comment (Konstantinov-Innokentii) [reply cmd/mimir/config-descriptor.json]: Thanks!

File: operations/mimir/mimir-flags-defaults.json
Comment (Konstantinov-Innokentii) [reply operations/mimir/mimir-flags-defaults.json:1]: Actually it was already added to main, so this diff is a github glitch - https://github.com/grafana/mimir/commit/23b6a82e7e1ad2baeadcff0cdb9324f395216794#diff-572ec83221bd126ed9d442b6f3b75989c37772084f29bde1134cf58b483e66d4R383

**Comment by Konstantinov-Innokentii** [comment]
@charleskorn Is there anything you want me to adress before merging?

File: pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go (hunk 10 above)
Comment (charleskorn) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go:574]: For `types.EnableManglingReturnedSlices = true`: we should probably set this in `init()` for each test package beneath `pkg/streamingpromql`.

For the other two: I think we should just enable and restore them in the places that need them (otherwise this might affect other tests).

Comment (Konstantinov-Innokentii) [reply pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go:574]: Ok, I'll address it in the sepate PR

**Comment by charleskorn** [comment]
> @charleskorn Is there anything you want me to adress before merging?

LGTM, happy for the remaining comment thread to be resolved in a follow-up PR
